### PR TITLE
v3 - complete revamp

### PR DESCRIPTION
**Pull request by brianjquinn**

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -93,29 +99,45 @@ class GraphQLComponent {
...
+      // construct a "gateway" schema from the schemas of imported
+      // components and this component's types/resolvers (if present)
+      this._schema = stitchSchemas({
+        subschemas,
+        typeDefs: this._types,
+        resolvers: this._resolvers,
+        schemaDirectives: this._directives
+      });
+    }
+    else {
```
Comment (mhvenkat) [inline lib/index.js:121]: Does this mean if we have imports, we don't need to create a federated schema even though the federated option is set to true in the individual GraphQL Components?

**Review by mhvenkat — COMMENTED** [review]
Does this mean if we have imports, we don't need to create a federated schema even though the federated option is set to true in the individual GraphQL Components?

File: examples/composition/listing-component/resolvers.js [hunk 2]
```diff
@@ -1,19 +1,33 @@
 'use strict';
 
+const GraphQLComponent = require('../../../lib');
+
 const resolvers = {
   Query: {
-    async listing(_, { id }, { dataSources }) {
-      const [property, reviews] = await Promise.all([
```
Comment (tlivings) [inline examples/composition/listing-component/resolvers.js:6]: Would be good to still have an example of utilizing the data source.

**Review by tlivings — COMMENTED** [review]
Concerned about removing memoize, which is a necessary performance enhacement to queries that may resolve the exact same sub-query multiple times.

File: lib/index.js (hunk 1 above)
Comment (brianjquinn) [reply lib/index.js:121]: yep - agree with this

File: examples/composition/listing-component/resolvers.js (hunk 2 above)
Comment (brianjquinn) [reply examples/composition/listing-component/resolvers.js:6]: technically this example does utilize a datasource - the top level component `listing-component` delegates down to reviews and property who make use of their own datasources

File: lib/resolvers/index.js [hunk 3]
```diff
@@ -2,194 +2,44 @@
...
- * wrapped/memoized
- * @param {string} fieldName -  the field on the parentType whose resolver
- * function is being wrapped/memoized
- * @param {function} resolve - the resolver function that parentType.
- * fieldName is mapped to
- * @returns {function} a function that wraps the input resolver function and
- * whose closure scope contains a WeakMap to achieve memoization of the wrapped
- * input resolver function
- */
-const memoize = function (parentType, fieldName, resolve) {
```
[the remark this answers is not in the document]
Comment (brianjquinn) [reply lib/resolvers/index.js:19]: my thinking was for simplicity - resolver wrapping has bitten us and I wanted to remove any potential for bugs considering this PR more or less utilizing graphql-tools code base to do heavy lifting

File: lib/index.js (hunk 1 above)
Comment (brianjquinn) [reply lib/index.js:121]: implemented this in my latest push

**Pull request merged**